x-pack/plugins/security/server/authentication/providers/kerberos.ts

```diff
@@ -0,0 +1,271 @@
+    } catch (err) {
+      this.logger.debug(`Failed to authenticate request via state: ${err.message}`);
+      return AuthenticationResult.failed(err);
+    }
+  }
+
+  /**
+   * This method is only called when authentication via access token stored in the state failed because of expired
+   * token. So we should use refresh token, that is also stored in the state, to extend expired access token and
+   * authenticate user with it.
+   * @param request Request instance.
+   * @param state State value previously stored by the provider.
+   */
+  private async authenticateViaRefreshToken(request: KibanaRequest, state: ProviderState) {
+    this.logger.debug('Trying to refresh access token.');
+
+    let refreshedTokenPair: TokenPair | null;
+    try {
+      refreshedTokenPair = await this.options.tokens.refresh(state.refreshToken);
+    } catch (err) {
+      return AuthenticationResult.failed(err);
+    }
+
+    // If refresh token is no longer valid, then we should clear session and renegotiate using SPNEGO.
+    if (refreshedTokenPair === null) {
+      this.logger.debug(
+        'Both access and refresh tokens are expired. Re-initiating SPNEGO handshake.'
+      );
+      return this.authenticateViaSPNEGO(request, state);
+    }
+
+    try {
+      const authHeaders = { authorization: `Bearer ${refreshedTokenPair.accessToken}` };
+      const user = await this.getUser(request, authHeaders);
+
+      this.logger.debug('Request has been authenticated via refreshed token.');
+      return AuthenticationResult.succeeded(user, { authHeaders, state: refreshedTokenPair });
+    } catch (err) {
+      this.logger.debug(
+        `Failed to authenticate user using newly refreshed access token: ${err.message}`
+      );
+      return AuthenticationResult.failed(err);
+    }
+  }
+
+  /**
+   * Tries to query Elasticsearch and see if we can rely on SPNEGO to authenticate user.
+   * @param request Request instance.
+   * @param [state] Optional state object associated with the provider.
+   */
+  private async authenticateViaSPNEGO(request: KibanaRequest, state?: ProviderState | null) {
+    this.logger.debug('Trying to authenticate request via SPNEGO.');
+
+    // Try to authenticate current request with Elasticsearch to see whether it supports SPNEGO.
+    let authenticationError: Error;
+    try {
+      await this.getUser(request);
+      this.logger.debug('Request was not supposed to be authenticated, ignoring result.');
+      return AuthenticationResult.notHandled();
+    } catch (err) {
+      // Fail immediately if we get unexpected error (e.g. ES isn't available). We should not touch
+      // session cookie in this case.
+      if (getErrorStatusCode(err) !== 401) {
+        return AuthenticationResult.failed(err);
+      }
+
+      authenticationError = err;
+    }
+
+    const challenges = ([] as string[]).concat(
+      get<string | string[]>(authenticationError, 'output.headers[WWW-Authenticate]') || ''
```

Comment:
I don't love that we're exposing Boom but I can also see the argument that this will make migrating easier.

I think when we add the new elasticsearch-js client (#35508), we should not use Boom.

Reply:
I'd suggest to introduce `ESClientError` type, which is `Boom` object atm. later we can switch all such places to a custom error and have more or less clear understanding what should be fixed